### PR TITLE
[SPARK-32540][SQL] Eliminate the filter clause in aggregate

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Spark SQL supported filter clause in aggregate, for example:
`select sum(distinct id) filter (where sex = 'man') from student;`
But sometimes we can eliminate the filter clause in aggregate.
`SELECT COUNT(DISTINCT 1) FILTER (WHERE true) FROM testData;`
could be transformed to
`SELECT COUNT(DISTINCT 1) FROM testData;`
`SELECT COUNT(DISTINCT 1) FILTER (WHERE false) FROM testData;`
could be transformed to
`SELECT 0 FROM testData;`

### Why are the changes needed?
Optimize the filter clause in aggregation

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
New test.

### CODE REVIEW

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -339,6 +340,39 @@ object EliminateDistinct extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Remove useless FILTER clause for aggregate expressions.
+ * This rule should be applied before RewriteDistinctAggregates.
+ */
+object EliminateAggregateFilter extends Rule[LogicalPlan] {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:347]: Can we put this rule in the main optimizer batch?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -339,6 +340,39 @@ object EliminateDistinct extends Rule[LogicalPlan] {
...
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformExpressions  {
+    case ae @ AggregateExpression(_, _, _, Some(Literal.TrueLiteral), _) =>
+      ae.copy(filter = None)
+    case ae @ AggregateExpression(af, _, _, Some(Literal.FalseLiteral), _) =>
+      rewrite(ae, af)
+    case ae @ AggregateExpression(af, _, _, Some(e), _)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:353]: We don't need to handle this. We should assume `ConstantFolding` will take care of it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -339,6 +340,39 @@ object EliminateDistinct extends Rule[LogicalPlan] {
...
+      if e.foldable && !ConvertToLocalRelation.hasUnevaluableExpr(e) =>
+      Literal.create(e.eval(EmptyRow), e.dataType) match {
+        case Literal.TrueLiteral => ae.copy(filter = None)
+        case Literal.FalseLiteral => rewrite(ae, af)
+        case _ => ae
+      }
+  }
+
+  private def rewrite(ae: AggregateExpression, af: AggregateFunction): Expression = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:362]: A more general way is to evaluate the agg func without any input.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:347]: OK.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: OK

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -339,6 +341,28 @@ object EliminateDistinct extends Rule[LogicalPlan] {
...
+ */
+object EliminateAggregateFilter extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformExpressions  {
+    case ae @ AggregateExpression(_, _, _, Some(Literal.TrueLiteral), _) =>
+      ae.copy(filter = None)
+    case AggregateExpression(af: DeclarativeAggregate, _, _, Some(Literal.FalseLiteral), _) =>
+      val initializer = SafeProjection.create(af.initialValues)
+      val evaluator = SafeProjection.create(af.evaluateExpression :: Nil, af.aggBufferAttributes)
+      val buffer = new SpecificInternalRow(af.aggBufferAttributes.map(_.dataType))
+      val initBuffer = initializer(buffer).copy()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:356]: why do we need to copy?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:356]: Looking at `AggregationIterator.initializeBuffer`, I think we should do
```
val initProj = MutableProjection.create...
val buffer = new SpecificInternalRow(af.aggBufferAttributes.map(_.dataType))
initProj.target(buffer)(EmptyRow)
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  val testRelation = LocalRelation('a.int)
+
+  test("Eliminate Filter always is true") {
+    val query = testRelation
+      .select(sumDistinct('a, Some(Literal.TrueLiteral)).as('result))
+      .analyze
+    val answer = testRelation
+      .select(sumDistinct('a).as('result))
+      .analyze
+    assert(query != answer)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala:42]: This is not needed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  test("Eliminate Filter is foldable and always is true") {
+    val query = testRelation
+      .select(countDistinctWithFilter(GreaterThan(Literal(2), Literal(1)), 'a).as('result))
+      .analyze
+    val answer = testRelation
+      .select(countDistinct('a).as('result))
+      .analyze
+    assert(query != answer)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala:53]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:356]: no need copy.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala (hunk 5 above)
Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala:42]: OK

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala (hunk 6 above)
Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateAggregateFilterSuite.scala:53]: OK

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by beliefer** [comment]
@cloud-fan Thanks for your help!

**Pull request closed without merging**